[Site Isolation] Screen Orientation API

stwrt opened the pull request:
#### de041ea9a3c3d4078c606f83b0f52dae426e5e4e
<pre>
[Site Isolation] Screen Orientation API
<a href="https://bugs.webkit.org/show_bug.cgi?id=300948">https://bugs.webkit.org/show_bug.cgi?id=300948</a>
<a href="https://rdar.apple.com/156027815">rdar://156027815</a>

Reviewed by Sihui Liu.

Create test to verify screen orientation API in site isolation mode.

On iOS platform we test the readonly attributes along with on change events.

The on change event does not fire on the Mac as we seem to be always in landscape-primary mode.

The lock() and unlock() methods cannot be tested at this time due to underlying issues outside of site isolation.

* LayoutTests/http/tests/site-isolation/resources/screen-orientation-rotate-device.html: Added.
* LayoutTests/http/tests/site-isolation/screen-orientation-api-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/screen-orientation-api.html: Added.
* LayoutTests/platform/mac/http/tests/site-isolation/screen-orientation-api-expected.txt: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.cpp:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::setCurrentOrientation):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/RemotePageScreenOrientationManagerProxy.cpp: Added.
(WebKit::RemotePageScreenOrientationManagerProxy::create):
(WebKit::RemotePageScreenOrientationManagerProxy::RemotePageScreenOrientationManagerProxy):
(WebKit::RemotePageScreenOrientationManagerProxy::~RemotePageScreenOrientationManagerProxy):
(WebKit::RemotePageScreenOrientationManagerProxy::didReceiveMessage):
(WebKit::RemotePageScreenOrientationManagerProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageScreenOrientationManagerProxy.h: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.cpp.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::~WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::setCurrentOrientation):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::setDeviceOrientation):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/302333@main">https://commits.webkit.org/302333@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f13511f69b63f0e2bea449006c47708c753019ed

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/128772 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1029 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39602 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136154 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80147 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a9e7ec7b-6e3e-4c92-a63e-15c04f4c1535) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/980 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/906 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/98031 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/65935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/20378e80-e764-433a-b892-0566785a065e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131719 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/731 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115356 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/78639 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c3bdbdb5-1f9b-4ee1-9ae0-4a101515ab17) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/667 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33470 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79434 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109102 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/33963 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138613 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/846 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/826 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106572 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/899 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111694 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106379 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27088 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/689 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30225 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53261 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/915 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64203 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/765 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/822 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/857 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/http/tests/site-isolation/screen-orientation-api-expected.txt:
```diff
@@ -0,0 +1,7 @@
+CONSOLE MESSAGE: onchange event fired
+PASS successfullyParsed is true
+
+TEST COMPLETE
```
When `TEST COMPLETE` is printed before the test is actually complete, this is an indication the test is not correctly written.

cdumez commented on LayoutTests/http/tests/site-isolation/screen-orientation-api.html:
```diff
@@ -0,0 +1,16 @@
+<script src="/js-test-resources/js-test.js"></script>
+<script>
+    if (window.testRunner) {
+        testRunner.dumpAsText();
```
Not needed when using `js-test.js`.

cdumez commented on LayoutTests/http/tests/site-isolation/screen-orientation-api.html:
```diff
@@ -0,0 +1,16 @@
+<script>
+    if (window.testRunner) {
+        testRunner.dumpAsText();
+        testRunner.waitUntilDone();
```
Should use `jsTestIsAsync = true;` instead when using `js-test.js`.

cdumez commented on LayoutTests/http/tests/site-isolation/screen-orientation-api.html:
```diff
@@ -0,0 +1,16 @@
+    addEventListener("message", (event) => {
+        testPassed("Received orientation message: " + event.data);
+        if (window.testRunner)
+            testRunner.notifyDone();
```
Should be using `finishJSTest();` instead when using `js-test.js`.

cdumez commented on LayoutTests/http/tests/site-isolation/screen-orientation-api.html:
```diff
@@ -0,0 +1,16 @@
+        testRunner.dumpAsText();
+        testRunner.waitUntilDone();
+    }
+
```
Please add a `description()` call since you're using `js-test.js`.

cdumez commented on LayoutTests/http/tests/site-isolation/screen-orientation-api.html:
```diff
@@ -0,0 +1,16 @@
+<!-- webkit-test-runner [ SiteIsolationEnabled=true ] -->
```
Does the test run in a *third-party* iframe simply because it lives in `http/tests/site-isolation`? I don't see the test doing anything specific to try and get a third-party iframe and thus leverage site isolation. I am not super familiar with how we do testing for site isolation.

szewai replied:
No: `http/tests/site-isolation` means the test will be run with 127.0.0.1; this header means to turn on site isolation for the test; and the test needs to cross-site content to test relevant behavior.
i.e. the test needs to create a cross-site frame and invoke Screen Orientation API in the frame.

szewai commented on Source/WebKit/UIProcess/RemotePageProxy.h:
```diff
@@ -130,6 +134,7 @@ class RemotePageProxy : public IPC::MessageReceiver, public RefCounted<RemotePag
     std::unique_ptr<RemotePageVisitedLinkStoreRegistration> m_visitedLinkStoreRegistration;
     WebPageProxyMessageReceiverRegistration m_messageReceiverRegistration;
     WebCore::MediaProducerMediaStateFlags m_mediaState;
+    RefPtr<WebScreenOrientationManagerProxy> m_screenOrientationManager;
```
Can we do something like https://commits.webkit.org/288830@main to introduce `RemotePageScreenOrientationManagerProxy` and have `RemotePageScreenOrientationManagerProxy` forward message to `WebScreenOrientationManagerProxy`?
Then we don't need to store process, page and remote page in `WebScreenOrientationManagerProxy`

stwrt replied:
Sure, let me look into that.

szewai commented on Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
```diff
@@ -333,14 +335,23 @@ static inline float adjustedUnexposedMaxEdge(float documentEdge, float exposedRe
+
+    // Update device orientation for main frame process
+    if (hasRunningProcess())
+        m_legacyMainFrameProcess->send(Messages::WebPage::SetDeviceOrientation(deviceOrientation), webPageIDInMainFrameProcess());
```
We have a helper function in `WebPageProxy`:
```
forEachWebContentProcess([](auto& process, auto pageID) {
        if (!process.hasConnection())
            return;
        process.send(Messages::WebPage::SetDeviceOrientation(deviceOrientation), pageID);
});
```

szewai commented on Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
```diff
@@ -49,14 +49,15 @@ WebScreenOrientationManagerProxy::WebScreenOrientationManagerProxy(WebPageProxy&
     unlockIfNecessary();
 
-    protectedPage()->protectedLegacyMainFrameProcess()->removeMessageReceiver(Messages::WebScreenOrientationManagerProxy::messageReceiverName(), m_page->webPageIDInMainFrameProcess());
+    if (RefPtr page = m_page.get())
```
`WebPageProxy` owns `WebScreenOrientationManagerProxy`, so I think we should not expect `m_page` to be null. Can we use `Ref` instead?

szewai commented on Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
```diff
@@ -78,7 +79,11 @@ void WebScreenOrientationManagerProxy::setCurrentOrientation(WebCore::ScreenOrie
         return;
 
-    protectedPage()->protectedLegacyMainFrameProcess()->send(Messages::WebScreenOrientationManager::OrientationDidChange(orientation), m_page->webPageIDInMainFrameProcess());
+    if (RefPtr page = m_page.get()) {
```
I think the change needed is:
```
Ref protectedPage { m_page.get() };
protectedPage->send(Messages::WebScreenOrientationManager::OrientationDidChange(orientation), protectedPage->webPageIDInMainFrameProcess());
```

szewai commented on Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
```diff
@@ -130,7 +135,9 @@ void WebScreenOrientationManagerProxy::lock(WebCore::ScreenOrientationLockType l
     if (resolvedLockedOrientation != m_currentlyLockedOrientation) {
         bool didLockOrientation = false;
-        Ref page = m_page.get();
+        RefPtr page = m_page.get();
```
We should convert `WeakRef` to `Ref` and the null check is unnecessary

szewai commented on Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
```diff
@@ -159,15 +166,17 @@ void WebScreenOrientationManagerProxy::unlock()
 
     bool didUnlockOrientation = false;
-    Ref page = m_page.get();
+    RefPtr page = m_page.get();
```
The null check is unnecessary

szewai commented on Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
```diff
@@ -78,7 +79,11 @@ void WebScreenOrientationManagerProxy::setCurrentOrientation(WebCore::ScreenOrie
+    if (RefPtr page = m_page.get()) {
+        Ref process = page->legacyMainFrameProcess();
+        process->send(Messages::WebScreenOrientationManager::OrientationDidChange(orientation), page->webPageIDInMainFrameProcess());
+    }
```
Shouldn't the message be sent to all web processes (remote page's process)?

szewai reviewed (approved):
r=me if bots are happy

stwrt committed:
[Site Isolation] Screen Orientation API
https://bugs.webkit.org/show_bug.cgi?id=300948
rdar://156027815

Reviewed by Sihui Liu.

Create test to verify screen orientation API in site isolation mode.

On iOS platform we test the readonly attributes along with on change events.

The on change event does not fire on the Mac as we seem to be always in landscape-primary mode.

The lock() and unlock() methods cannot be tested at this time due to underlying issues outside of site isolation.

* LayoutTests/http/tests/site-isolation/resources/screen-orientation-rotate-device.html: Added.
* LayoutTests/http/tests/site-isolation/screen-orientation-api-expected.txt: Added.
* LayoutTests/http/tests/site-isolation/screen-orientation-api.html: Added.
* LayoutTests/platform/mac/http/tests/site-isolation/screen-orientation-api-expected.txt: Added.
* Source/WebKit/Sources.txt:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.cpp:
* Source/WebKit/UIProcess/RemotePageProxy.cpp:
(WebKit::RemotePageProxy::injectPageIntoNewProcess):
(WebKit::RemotePageProxy::setCurrentOrientation):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/RemotePageScreenOrientationManagerProxy.cpp: Added.
(WebKit::RemotePageScreenOrientationManagerProxy::create):
(WebKit::RemotePageScreenOrientationManagerProxy::RemotePageScreenOrientationManagerProxy):
(WebKit::RemotePageScreenOrientationManagerProxy::~RemotePageScreenOrientationManagerProxy):
(WebKit::RemotePageScreenOrientationManagerProxy::didReceiveMessage):
(WebKit::RemotePageScreenOrientationManagerProxy::didReceiveSyncMessage):
* Source/WebKit/UIProcess/RemotePageScreenOrientationManagerProxy.h: Copied from Source/WebKit/UIProcess/RemoteLayerTree/RemoteMonotonicTimeline.cpp.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.cpp:
(WebKit::WebScreenOrientationManagerProxy::WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::~WebScreenOrientationManagerProxy):
(WebKit::WebScreenOrientationManagerProxy::setCurrentOrientation):
* Source/WebKit/UIProcess/WebScreenOrientationManagerProxy.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::setDeviceOrientation):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/302333@main
